### PR TITLE
[WIP] RESTful Actions

### PR DESCRIPTION
**Pull request by mateusg**
Discussion started at #32 

``` ruby
class MyApiResource
  include Jeckle::Resource
  api :some_api
  rest_actions except: [:delete]
end
```
## TO DO
- [ ] Create rest_actions into module HTTP
- [ ] Add options only and except
- [ ] .find => GET /posts/:id
- [ ] .create => POST /posts
- [ ] #save => POST /posts 
- [ ] #update => PATCH /posts 
- [ ] #destroy => DELETE /posts/:id

> This branch was forked from 0.4.0.

### CODE REVIEW

**Comment by mateusg** [comment]
@tomas-stefano Can I delete branch `0.4.0`? Everything it has is now here, I think we won't need it anymore.

**Comment by tomas-stefano** [comment]
@mateusg 
yesyoucan

File: .travis.yml [hunk 1]
```diff
@@ -7,6 +7,7 @@ rvm:
   - ruby-head
   - ruby
   - jruby
+  - 2.2.2
   - 2.1.0
   - 2.0.0
   - 1.9.3
```
Comment (tomas-stefano) [inline .travis.yml]: It's not related, but can we remove support from 1.9.3?

**Comment by mateusg** [comment]
@tomas-stefano hahaha, ok! done.

File: spec/fixtures/resources/comment_legacy.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+class CommentLegacy
+  include Jeckle::Resource
+
+  api :my_super_api
+  root collection: 'super-comments', member: true
```
Comment (tomas-stefano) [inline spec/fixtures/resources/comment_legacy.rb]: My super comment: :+1:

**Comment by tomas-stefano** [comment]
:+1: :ship: after CI ok.

File: .travis.yml (hunk 1 above)
Comment (tomas-stefano) [inline .travis.yml]: I will do in a diff PR this removal.

**Comment by mateusg** [comment]
@tomas-stefano I still have many things to do here, please don't merge it yet. :)

**Comment by tomas-stefano** [comment]
Ok. :+1: Tell me when you're ready and I will finish the review and merge.

File: .travis.yml (hunk 1 above)
Comment (mateusg) [inline .travis.yml]: I'm ok with or without it. @brennovich ?

Comment (brennovich) [inline .travis.yml]: Bye `1.9.3` o/

**Comment by mateusg** [comment]
@tomas-stefano @brennovich Guys, I've been brainstorming about this feature and I came to some conclusions.

At first I was thinking about this `restful_actions` DSL as a way to define routes/urls/endpoints for the usual RESTful methods (`index`, `show`, `create`, `update`, `destroy`), just like the `resources` DSL defines Rails routes. So, following this thought, these are the methods that would be defined/exposed for each action:

Action                | Defines Class Methods          |   Defines Instance Methods
--------------------  | --------------------------------------- | ---------------------------------- 
index                 | search
show                 | find
create               | create                                      | save 
update              | update                                     | save
destroy             | destroy_by_id                          | destroy

What is the problem with this? It requires the user to do this configuration for **every** `Jeckle::Resource` he creates. If he doesn't add `restful_actions :index` to his resource, he won't even be able to do `MyResource.search`.

So what do you guys think about abandoning this DSL and just create all RESTful actions for every resource? I think those methods should be defined and working by default, and if the user needs custom behaviour, he will have to use the [custom action DSL](https://github.com/tomas-stefano/jeckle/issues/35), like so:

```ruby
action :create, method: :post, serializer: ShotCreateSerializer
```
That would overwrite the default `create` method.

**Comment by brennovich** [comment]
@mateusg IMO there is no problem to force the user to call `restful_actions` macro. We can make it to include all actions by default just for calling `restful_actions`.

**Comment by brennovich** [comment]
@mateusg :+1:

**Comment by mateusg** [comment]
Me and @brennovich talked a little and we think the `restful_actions` should use the `action` DSL.

```ruby
def restful_actions(actions, options)
          # /api/v1/resources/
          action :search, method: :get, path: '/', on: :collection

          # /api/v1/resources/:id/
          action :find, method: :get, path: '/', on: :member

          # /api/v1/resources/create
          action :create, method: :post

          # /api/v1/resources/:id/destroy
          action :destroy, method: :delete

          # /api/v1/resources/:id/update
          action :update, method: :put
        end
```

We still don't know how `save` will behave. I'll focus on developing the action DSL and adding `search` and `find` methods. We'll see the rest later.

PS: There will be no `index` and `show` actions. They will be called `search` and `find` instead.

**Comment by tomas-stefano** [comment]
:+1:

**Pull request closed without merging**